homeassistant/components/fritz/switch.py

```diff
@@ -0,0 +1,522 @@
+"""Switches for AVM Fritz!Box functions."""
+from __future__ import annotations
+
+from collections import OrderedDict
+import logging
+from typing import Any
+
+from fritzconnection.core.exceptions import FritzConnectionException, FritzSecurityError
+import xmltodict
+
+from homeassistant.components.network.util import async_get_source_ip
+from homeassistant.components.switch import SwitchEntity
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .common import FritzBoxBaseSwitch, FritzBoxTools, SwitchInfo
+from .const import (
+    DOMAIN,
+    SWITCH_TYPE_DEFLECTION,
+    SWITCH_TYPE_DEVICEPROFILE,
+    SWITCH_TYPE_PORTFORWARD,
+    SWITCH_TYPE_WIFINETWORK,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def service_call_action(
+    fritzbox_tools: FritzBoxTools,
+    service_name: str,
+    service_suffix: str | None,
+    action_name: str,
+    **kwargs: Any,
+) -> None | dict:
+    """Return service details."""
+
+    if f"{service_name}{service_suffix}" not in fritzbox_tools.connection.services:
+        return None
+
+    try:
+        return await fritzbox_tools.hass.async_add_executor_job(
+            lambda: fritzbox_tools.connection.call_action(
+                f"{service_name}:{service_suffix}", action_name, **kwargs
+            )
+        )
+    except FritzSecurityError:
+        _LOGGER.error(
+            "Authorization Error: Please check the provided credentials and verify that you can log into the web interface",
+            exc_info=True,
+        )
+        return None
+    except FritzConnectionException:
+        _LOGGER.error(
+            "Connection Error: Please check the device is properly configured for remote login",
+            exc_info=True,
+        )
+        return None
+    except Exception:  # pylint: disable=broad-except
+        _LOGGER.error(
+            "Service/Action Error: cannot execute service %s",
+            service_name,
+            exc_info=True,
+        )
+        return None
+
+
+async def get_deflections(
+    fritzbox_tools: FritzBoxTools, service_name: str
+) -> list[OrderedDict[Any, Any]] | None:
+    """Get deflection switch info."""
+
+    deflection_list = await service_call_action(
+        fritzbox_tools,
+        service_name,
+        "1",
+        "GetDeflections",
+    )
+
+    if not deflection_list:
+        return []
+
+    deflections: list = []
+    deflections.append(
+        xmltodict.parse(deflection_list["NewDeflectionList"])["List"]["Item"]
+    )
+
+    return deflections
+
+
+async def get_entities_list(
+    fritzbox_tools: FritzBoxTools, device_friendly_name: str
+) -> list:
+    """Get a list of all switches to create."""
+    entities_list: list[
+        FritzBoxDeflectionSwitch
+        | FritzBoxPortSwitch
+        | FritzBoxProfileSwitch
+        | FritzBoxWifiSwitch
+    ] = []
+
+    # 1. Deflection switches
```

Comment:
This would be easier to follow if 1,2,3,4 were their own functions.

Reply:
Done